### ISSUE TITLE
Support for packages that depend on public sublibraries

### ISSUE THREAD

**Issue by dcastro** [commercialhaskell/stack#5318]
Seems like stack does not currently (v2.3.1) support projects with multiple packages, where:
* one package exposes multiple public libraries
* and another package depends on those public sublibraries.

I've create a small repro to show what I'm trying to achieve [here](https://github.com/dcastro/demo233/tree/multi-lib) (branch: multi-lib).
`stack build pkg-b` fails with:

<details>
<summary>stack log</summary>

```

SubLibrary dependency is not supported, this will almost certainly fail
pkg-b> configure (lib)
Configuring pkg-b-0.1.0.0...
Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.3: Encountered missing or private
dependencies:
pkg-a : {pkg-a, pkg-a2} ==0.1.0.0


--  While building package pkg-b-0.1.0.0 using:
      /home/dc/.stack/setup-exe-cache/x86_64-linux-tinfo6/Cabal-simple_mPHDZzAJ_3.0.1.0_ghc-8.8.3 --builddir=.stack-work/dist/x86_64-linux-tinfo6/Cabal-3.0.1.0 configure --user --package-db=clear --package-db=global --package-db=/home/dc/.stack/snapshots/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/pkgdb --package-db=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/pkgdb --libdir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/lib --bindir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/bin --datadir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/share --libexecdir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/libexec --sysconfdir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/etc --docdir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/doc/pkg-b-0.1.0.0 --htmldir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/doc/pkg-b-0.1.0.0 --haddockdir=/home/dc/dev/serokell/demo233/.stack-work/install/x86_64-linux-tinfo6/a6119f4790a036ddc181a55d35e8fb09910df5235ac7a59872f78c70ded9011a/8.8.3/doc/pkg-b-0.1.0.0 --dependency=base=base-4.13.0.0 --dependency=pkg-a=pkg-a-0.1.0.0-80giyUDLKHPCSvIpWeSK7i --exact-configuration --ghc-option=-fhide-source-paths
    Process exited with code: ExitFailure 1

```

</details>

I noticed there's an [integration test](https://github.com/commercialhaskell/stack/tree/master/test/integration/tests/cabal-sublibrary-dependency) for this, which [is expected to fail](https://github.com/commercialhaskell/stack/blob/master/test/integration/tests/cabal-sublibrary-dependency/Main.hs#L8-L14).

I was wondering if this feature is in the roadmap (short/long-term)?

**Comment by amigalemming** [comment]
As a first measure it would be great to know which of the dependencies causes the problem.

**Comment by brandon-leapyear** [comment]
@amigalemming what do you mean? It seems like the error is just saying that the public sublibraries are "missing"
```
Encountered missing or private
dependencies:
pkg-a : {pkg-a, pkg-a2} ==0.1.0.0
```

@kk-hainq @philderbeast What is the status of your PRs? (https://github.com/commercialhaskell/stack/pull/5659, https://github.com/commercialhaskell/stack/pull/5839) Seems like it's blocked on Stack using Cabal 3.8 itself? @mpilgrem any updates on https://github.com/commercialhaskell/stack/pull/5887? Anything I could help with?

**Comment by amigalemming** [comment]
> @amigalemming what do you mean? It seems like the error is just saying that the public sublibraries are "missing"
> 
> ```
> Encountered missing or private
> dependencies:
> pkg-a : {pkg-a, pkg-a2} ==0.1.0.0
> ```

That seems to be new.

**Comment by brandonchinn178** [comment]
I just copied the message from the issue description

**Comment by mpilgrem** [comment]
@brandonchinn178, one of my priorities is moving Stack and its CI to GHC 9.4.3 and `Cabal-3.8.1.0`. I will return to #5887 soon - I am waiting for GitHub-hosted runners to move from Stack 2.9.1 to Stack 2.9.3 (which version of Stack should be able to build that version of Stack).

**Comment by mpilgrem** [comment]
I am going to close this issue as a duplicate of the earlier #5256.
